pkg/kv/kvserver/loqrecovery/server.go

```diff
@@ -204,6 +216,182 @@ func (s Server) ServeClusterReplicas(
 		})
 }
 
+func (s Server) StagePlan(
+	ctx context.Context, req *serverpb.RecoveryStagePlanRequest,
+) (*serverpb.RecoveryStagePlanResponse, error) {
+	if !req.ForcePlan && req.Plan == nil {
+		return nil, errors.New("stage plan request can't be used with empty plan without force flag")
+	}
+	localNodeID := s.nodeIDContainer.Get()
+	// Create a plan copy with all empty fields to shortcut all plan nil checks
+	// below to avoid unnecessary nil checks.
+	var plan loqrecoverypb.ReplicaUpdatePlan
+	if req.Plan != nil {
+		plan = *req.Plan
+	}
+	if req.AllNodes {
+		// Scan cluster for conflicting recovery plans and for stray nodes that are
+		// planned for forced decommission, but rejoined cluster.
+		foundNodes := make(map[roachpb.NodeID]struct{})
+		err := s.visitNodes(
+			ctx,
+			replicaInfoStreamRetryOptions,
+			func(nodeID roachpb.NodeID, client serverpb.AdminClient) error {
+				res, err := client.RecoveryNodeStatus(ctx, &serverpb.RecoveryNodeStatusRequest{})
+				if err != nil {
+					return errors.Mark(err, errMarkRetry)
+				}
+				// If operation fails here, we don't want to find all remaining
+				// violating nodes because cli must ensure that cluster is safe for
+				// staging.
+				if !req.ForcePlan && res.Status.PendingPlanID != nil && !res.Status.PendingPlanID.Equal(plan.PlanID) {
+					return errors.Newf("plan %s is already staged on node n%d", res.Status.PendingPlanID, nodeID)
+				}
+				foundNodes[nodeID] = struct{}{}
+				return nil
+			})
+		if err != nil {
+			return nil, err
+		}
+
+		// Check that no nodes that must be decommissioned are present.
+		for _, dID := range plan.DecommissionedNodeIDs {
+			if _, ok := foundNodes[dID]; ok {
+				return nil, errors.Newf("node n%d was planned for decommission, but is present in cluster", dID)
+			}
+		}
+
+		// Check out that all nodes that should save plan are present.
+		for _, u := range plan.Updates {
+			if _, ok := foundNodes[u.NodeID()]; !ok {
+				return nil, errors.Newf("node n%d has planned changed but is unreachable in the cluster", u.NodeID())
+			}
+		}
+
+		// Distribute plan - this should not use fan out to available, but use
```

Comment:
If we shouldn't use fanout, why don't we dial them directly? It doesn't seem like we check for unexpected nodes here either, i.e. nodes that are not in `foundNodes` but that are found via `visitNodes()`, so a node could come back online after the verification above.

Reply:
That's 	a fortunate or unfortunate leak of the comment from first iteration. I was thinking that we can use same set of nodes to distribute the plan, but it looks like we don't need to do that. We can do fan out again, since a this point it is not reversible anyways and once we started we should keep trying. The best thing we could do is report if plan didn't reach some nodes that we wanted it to reach.